### ISSUE TITLE
No internet connection

### ISSUE THREAD

**Issue by simondvt** [anbox/anbox#1873]
I have installed anbox following the [arch wiki](https://wiki.archlinux.org/title/Anbox), it works except for internet connection. In particular I did execute `nmcli con add type bridge ifname anbox0 -- connection.id anbox-net ipv4.method shared ipv4.addresses 192.168.250.1/24` since I use NetworkManager.
I run `sudo systemctl restart anbox-container-manager.service` after creating the anbox0 interface.

<details>
<summary>`ip a` in linux</summary>

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp59s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN group default qlen 1000
    link/ether 84:7b:eb:2d:62:30 brd ff:ff:ff:ff:ff:ff
3: wlp60s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 9c:b6:d0:0d:80:a3 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.238/24 brd 192.168.1.255 scope global dynamic noprefixroute wlp60s0
       valid_lft 18074sec preferred_lft 18074sec
    inet6 fe80::e3d2:a4fb:cf39:6086/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
4: anbox0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 1a:4f:15:fd:c5:f6 brd ff:ff:ff:ff:ff:ff
    inet 192.168.250.1/24 brd 192.168.250.255 scope global noprefixroute anbox0
       valid_lft forever preferred_lft forever
10: veth7wtoQC@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master anbox0 state UP group default qlen 1000
    link/ether fe:3d:59:bc:b6:d2 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fe80::fc3d:59ff:febc:b6d2/64 scope link 
       valid_lft forever preferred_lft forever
```
</details>

<details>
<summary>`ip a` in anbox</summary>

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0@if10: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 0a:da:05:a6:f8:f7 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 192.168.250.2/24 brd 192.168.250.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::8da:5ff:fea6:f8f7/64 scope link 
       valid_lft forever preferred_lft forever
```
</details>

<details>
<summary>anbox system-info output</summary>

```
version: local-04ac697
cpu:
  arch:  x86
  brand: Intel(R) Core(TM) i7-6820HK CPU @ 2.70GHz
  features: 
    - aes
    - sse4_1
    - sse4_2
    - avx
    - avx2
os:
  name: Arch Linux
  version: 
  snap-based: false
kernel:
  version: Linux version 5.13.4-zen2-1-zen (linux-zen@archlinux) (gcc (GCC) 11.1.0, GNU ld (GNU Binutils) 2.36.1) #1 ZEN SMP PREEMPT Thu, 22 Jul 2021 20:08:57 +0000
  binder: false
  binderfs: true
  ashmem: true
graphics:
  egl:
    vendor: Mesa Project
    version: 1.5
    extensions:
      - EGL_ANDROID_blob_cache
      - EGL_ANDROID_native_fence_sync
      - EGL_CHROMIUM_sync_control
      - EGL_EXT_buffer_age
      - EGL_EXT_create_context_robustness
      - EGL_EXT_image_dma_buf_import
      - EGL_EXT_image_dma_buf_import_modifiers
      - EGL_EXT_swap_buffers_with_damage
      - EGL_IMG_context_priority
      - EGL_KHR_cl_event2
      - EGL_KHR_config_attribs
      - EGL_KHR_create_context
      - EGL_KHR_create_context_no_error
      - EGL_KHR_fence_sync
      - EGL_KHR_get_all_proc_addresses
      - EGL_KHR_gl_colorspace
      - EGL_KHR_gl_renderbuffer_image
      - EGL_KHR_gl_texture_2D_image
      - EGL_KHR_gl_texture_3D_image
      - EGL_KHR_gl_texture_cubemap_image
      - EGL_KHR_image
      - EGL_KHR_image_base
      - EGL_KHR_image_pixmap
      - EGL_KHR_no_config_context
      - EGL_KHR_reusable_sync
      - EGL_KHR_surfaceless_context
      - EGL_KHR_swap_buffers_with_damage
      - EGL_EXT_pixel_format_float
      - EGL_KHR_wait_sync
      - EGL_MESA_configless_context
      - EGL_MESA_drm_image
      - EGL_MESA_image_dma_buf_export
      - EGL_MESA_query_driver
      - EGL_NOK_texture_from_pixmap
      - EGL_WL_bind_wayland_display
  gles2:
    vendor: Intel
    vendor: OpenGL ES 3.2 Mesa 21.1.5
    extensions:
      - GL_EXT_blend_minmax
      - GL_EXT_multi_draw_arrays
      - GL_EXT_texture_filter_anisotropic
      - GL_EXT_texture_compression_s3tc
      - GL_EXT_texture_compression_dxt1
      - GL_EXT_texture_compression_rgtc
      - GL_EXT_texture_format_BGRA8888
      - GL_OES_compressed_ETC1_RGB8_texture
      - GL_OES_depth24
      - GL_OES_element_index_uint
      - GL_OES_fbo_render_mipmap
      - GL_OES_mapbuffer
      - GL_OES_rgb8_rgba8
      - GL_OES_standard_derivatives
      - GL_OES_stencil8
      - GL_OES_texture_3D
      - GL_OES_texture_float
      - GL_OES_texture_float_linear
      - GL_OES_texture_half_float
      - GL_OES_texture_half_float_linear
      - GL_OES_texture_npot
      - GL_OES_vertex_half_float
      - GL_EXT_draw_instanced
      - GL_EXT_texture_sRGB_decode
      - GL_OES_EGL_image
      - GL_OES_depth_texture
      - GL_AMD_performance_monitor
      - GL_OES_packed_depth_stencil
      - GL_EXT_texture_type_2_10_10_10_REV
      - GL_NV_conditional_render
      - GL_OES_get_program_binary
      - GL_APPLE_texture_max_level
      - GL_EXT_discard_framebuffer
      - GL_EXT_read_format_bgra
      - GL_EXT_frag_depth
      - GL_NV_fbo_color_attachments
      - GL_OES_EGL_image_external
      - GL_OES_EGL_sync
      - GL_OES_vertex_array_object
      - GL_OES_viewport_array
      - GL_ANGLE_pack_reverse_row_order
      - GL_ANGLE_texture_compression_dxt3
      - GL_ANGLE_texture_compression_dxt5
      - GL_EXT_occlusion_query_boolean
      - GL_EXT_robustness
      - GL_EXT_texture_rg
      - GL_EXT_unpack_subimage
      - GL_NV_draw_buffers
      - GL_NV_read_buffer
      - GL_NV_read_depth
      - GL_NV_read_depth_stencil
      - GL_NV_read_stencil
      - GL_EXT_draw_buffers
      - GL_EXT_map_buffer_range
      - GL_KHR_debug
      - GL_KHR_robustness
      - GL_KHR_texture_compression_astc_ldr
      - GL_NV_pixel_buffer_object
      - GL_OES_depth_texture_cube_map
      - GL_OES_required_internalformat
      - GL_OES_surfaceless_context
      - GL_EXT_color_buffer_float
      - GL_EXT_sRGB_write_control
      - GL_EXT_separate_shader_objects
      - GL_EXT_shader_framebuffer_fetch
      - GL_EXT_shader_group_vote
      - GL_EXT_shader_implicit_conversions
      - GL_EXT_shader_integer_mix
      - GL_EXT_tessellation_point_size
      - GL_EXT_tessellation_shader
      - GL_INTEL_conservative_rasterization
      - GL_INTEL_performance_query
      - GL_ANDROID_extension_pack_es31a
      - GL_EXT_base_instance
      - GL_EXT_compressed_ETC1_RGB8_sub_texture
      - GL_EXT_copy_image
      - GL_EXT_draw_buffers_indexed
      - GL_EXT_draw_elements_base_vertex
      - GL_EXT_gpu_shader5
      - GL_EXT_polygon_offset_clamp
      - GL_EXT_primitive_bounding_box
      - GL_EXT_render_snorm
      - GL_EXT_shader_io_blocks
      - GL_EXT_texture_border_clamp
      - GL_EXT_texture_buffer
      - GL_EXT_texture_cube_map_array
      - GL_EXT_texture_norm16
      - GL_EXT_texture_view
      - GL_KHR_blend_equation_advanced
      - GL_KHR_blend_equation_advanced_coherent
      - GL_KHR_context_flush_control
      - GL_KHR_robust_buffer_access_behavior
      - GL_NV_image_formats
      - GL_OES_copy_image
      - GL_OES_draw_buffers_indexed
      - GL_OES_draw_elements_base_vertex
      - GL_OES_gpu_shader5
      - GL_OES_primitive_bounding_box
      - GL_OES_sample_shading
      - GL_OES_sample_variables
      - GL_OES_shader_io_blocks
      - GL_OES_shader_multisample_interpolation
      - GL_OES_tessellation_point_size
      - GL_OES_tessellation_shader
      - GL_OES_texture_border_clamp
      - GL_OES_texture_buffer
      - GL_OES_texture_cube_map_array
      - GL_OES_texture_stencil8
      - GL_OES_texture_storage_multisample_2d_array
      - GL_OES_texture_view
      - GL_EXT_blend_func_extended
      - GL_EXT_buffer_storage
      - GL_EXT_float_blend
      - GL_EXT_geometry_point_size
      - GL_EXT_geometry_shader
      - GL_EXT_shader_samples_identical
      - GL_EXT_texture_sRGB_R8
      - GL_KHR_no_error
      - GL_KHR_texture_compression_astc_sliced_3d
      - GL_NV_fragment_shader_interlock
      - GL_OES_EGL_image_external_essl3
      - GL_OES_geometry_point_size
      - GL_OES_geometry_shader
      - GL_OES_shader_image_atomic
      - GL_EXT_clip_cull_distance
      - GL_EXT_disjoint_timer_query
      - GL_EXT_texture_compression_s3tc_srgb
      - GL_MESA_shader_integer_functions
      - GL_EXT_clip_control
      - GL_EXT_color_buffer_half_float
      - GL_EXT_semaphore
      - GL_EXT_semaphore_fd
      - GL_EXT_texture_compression_bptc
      - GL_KHR_parallel_shader_compile
      - GL_EXT_EGL_image_storage
      - GL_EXT_shader_framebuffer_fetch_non_coherent
      - GL_EXT_texture_shadow_lod
      - GL_INTEL_blackhole_render
      - GL_MESA_framebuffer_flip_y
      - GL_NV_compute_shader_derivatives
      - GL_EXT_demote_to_helper_invocation
      - GL_EXT_depth_clamp
      - GL_EXT_texture_query_lod

```
</details>

The systemd log doesn't show any errors.
When i ping `192.168.250.1` inside anbox I get a response, however if I ping 8.8.8.8 I don't get any and I have 100% packet loss.
I checked [this](https://github.com/anbox/anbox/issues/443#issuecomment-345449671) but no luck.

**Comment by deakus** [comment]
same problem on Ubuntu 20.04, running kernel 5.4.0-80, binder kernel module is missing, should be in  kernel > 5.0 right ?
 
Running Anbox 4 in snap
Tried accessing anbox shell to set default route as found here

[anbox shell](https://gitlab.com/postmarketOS/pmaports/-/issues/480)

script can be found for at (in my installation)  /snap/anbox/186/bin/anbox-shell.sh

Howeve it fails as t is looking for a lxc container

**Comment by Fuseteam** [comment]
huh odd, when do see the anbox0 interface i do have internet access just fine. is the interface still there when you noticed it? if yes, perhaps you need to start anbox's bridge; the command for that escapes my tongue atm

**Comment by simondvt** [comment]
> huh odd, when do see the anbox0 interface i do have internet access just fine. is the interface still there when you noticed it? if yes, perhaps you need to start anbox's bridge; the command for that escapes my tongue atm

Thanks, running anbox-bridge fixed the issue!

**Comment by XploreSecF** [comment]
idk

**Comment by sim590** [comment]
I'm having the same issue...

My computer's `ip a` output:

<details>
<p>

```ip
11: anbox0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 06:bc:ac:26:32:bf brd ff:ff:ff:ff:ff:ff
    inet 192.168.250.1/24 scope global anbox0
       valid_lft forever preferred_lft forever
    inet6 fe80::4bc:acff:fe26:32bf/64 scope link
       valid_lft forever preferred_lft forever
12: vethpY3ABo@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master anbox0 state UP group default qlen 1000
    link/ether fe:78:5f:8c:54:6e brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fe80::fc78:5fff:fe8c:546e/64 scope link
       valid_lft forever preferred_lft forever

```

</p>
</details>

`ip a` from `adb shell`:
<details>

```ip
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether ce:da:2f:27:d5:82 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 192.168.250.1/24 brd 192.168.250.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::ccda:2fff:fe27:d582/64 scope link
       valid_lft forever preferred_lft forever
```

</details>

I have this service which runs `anbox-bridge.sh`:
<details>

```systemd
# /usr/lib/systemd/system/anbox-container-manager.service
[Unit]
Description=Anbox Container Manager

[Service]
ExecStart=/usr/bin/anbox container-manager --daemon --privileged --data-path=/var/lib/anbox

[Install]
WantedBy=multi-user.target
Also=dev-binderfs.mount

# /etc/systemd/system/anbox-container-manager.service.d/override.conf
[Service]
ExecStartPre=/usr/local/bin/anbox-bridge.sh start
ExecStart=
ExecStart=/usr/bin/anbox container-manager --daemon --privileged --data-path=/var/lib/anbox --container-network-address=192.168.250.1/24 --container-network-gateway=192.168.250.255
ExecStopPost=/usr/local/bin/anbox-bridge.sh stop
```

</details>

But as you can see, anbox-bridge doesn't solve my issue... What's wrong ?

**Comment by Fuseteam** [comment]
@sim590 anbox 0 being up implies that it should work, how are you testing for internet connectivity?

**Comment by sim590** [comment]
@Fuseteam I did `ping 8.8.8.8` and it doesn't give anything. Pinging, `192.168.250.1` works. Also, Google Play Store is in a loop on "Checking info...".

**Comment by Fuseteam** [comment]
where are you pinging `8.8.8.8`? what package format did you use for anbox?

**Comment by sim590** [comment]
>what package format did you use for anbox?

I used the Arch AUR packages and followed the instructions on the arch wiki.

>where are you pinging 8.8.8.8

I guess that is a Google DNS server or something. It always responds to ping normally.

**Comment by Fuseteam** [comment]
> I guess that is a Google DNS server or something. It always responds to ping normally.

i mean where are you pinging from? from side anbox or from you pc?

> I used the Arch AUR packages and followed the instructions on the arch wiki.

hmm that doesn't sound like snap, i have no idea how to debug this

**Comment by SpacingBat3** [comment]
@sim590 @Fuseteam I did a small research about what could cause the issues with Anbox bridge under Arch Linux and found that containers using bridges like Docker or Anbox may be broken with the `nftables` (see [nftables#Working_with_Docker][wiki]):

> Using nftables can interfere with Docker networking (and probably other container runtimes as well).

The similar issue did happen for me as well – I had a bridge set up, but there was no internet connection. As I found out that `nftables` may be actually a cause to that, I removed it and replaced `iptables-nft` with *legacy* `iptables`. After a reboot, everything seemed to work as intended, including other bridges I had set up in my OS.

[wiki]: https://wiki.archlinux.org/title/Nftables#Working_with_Docker "nftables – ArchWiki"

**Comment by Fuseteam** [comment]
oh that's an interesting find, i wonder if there's a way to make it work with nftables

**Comment by sim590** [comment]
@SpacingBat3 is that relevant to my case if I have `iptables` package installed?
```
$ pacman -Qs iptables
local/iptables 1:1.8.7-1
    Linux kernel packet control tool (using legacy interface)
```
and `nftables` doesn't give any result, so it seems like it's not installed on my system?

**Comment by SpacingBat3** [comment]
> @SpacingBat3 is that relevant to my case if I have `iptables` package installed?
> 
> ```
> $ pacman -Qs iptables
> local/iptables 1:1.8.7-1
>     Linux kernel packet control tool (using legacy interface)
> ```
> 
> and `nftables` doesn't give any result, so it seems like it's not installed on my system?

Yes, you seem to have a legacy `iptables` package installed... Other ideas came to my mind that could be an issue:
 - Is `iptables.service` running? Are there any rules set for it (`iptables --list-rules`)?
 - Are there any *conflicting* packages? For instance, I'm unsure if you can use `ufw` firewall (which seems to be installed in Manjaro by default) with Anbox, as it can conflict with `iptables.service` and I don't know how much it interference with.
 -  Are you connected through IPv4 or IPv6? For IPv6, an additional configuration might be needed for NAT to work (if Anbox is capable of working with IPv6), like in case of Docker...
 
 There's a lot that could break in process. If none of these will help you, check your bridge configuration again just to make sure it is configured as in Arch Wiki.

**Comment by sim590** [comment]
@SpacingBat3:

TLDR: I did remove these parameters (which I added manually) from Systemd `ExecStart` line:

```
--container-network-address=192.168.250.1/24 --container-network-gateway=192.168.250.255
```

while I was trying to debug the issue. It now works.

I did find that out that while responding to your message, so I did leave the response below anyway.

> * Is `iptables.service` running? Are there any rules set for it (`iptables --list-rules`)?

Here's the output:

<details>

```
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
-A INPUT -i anbox0 -p tcp -m tcp --dport 53 -m comment --comment "managed by anbox-bridge" -j ACCEPT
-A INPUT -i anbox0 -p udp -m udp --dport 53 -m comment --comment "managed by anbox-bridge" -j ACCEPT
-A INPUT -i anbox0 -p tcp -m tcp --dport 67 -m comment --comment "managed by anbox-bridge" -j ACCEPT
-A INPUT -i anbox0 -p udp -m udp --dport 67 -m comment --comment "managed by anbox-bridge" -j ACCEPT
-A FORWARD -o anbox0 -m comment --comment "managed by anbox-bridge" -j ACCEPT
-A FORWARD -i anbox0 -m comment --comment "managed by anbox-bridge" -j ACCEPT
```

</details>

> * Are there any _conflicting_ packages? For instance, I'm unsure if you can use `ufw` firewall (which seems to be installed in Manjaro by default) with Anbox, as it can conflict with `iptables.service` and I don't know how much it interference with.

I don't think so. The `ufw` package is not installed.

> * Are you connected through IPv4 or IPv6? For IPv6, an additional configuration might be needed for NAT to work (if Anbox is capable of working with IPv6), like in case of Docker...

Here's the output of `ip a` on my host computer:

<details>

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp7s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether b4:2e:99:f1:ec:59 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.101/24 brd 192.168.1.255 scope global dynamic noprefixroute enp7s0
       valid_lft 5881sec preferred_lft 5881sec
    inet6 fe80::9b9a:c20e:3396:4607/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
4: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:ea:28:c8:d6 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
7: anbox0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 06:bc:ac:26:32:bf brd ff:ff:ff:ff:ff:ff
    inet 192.168.250.1/24 scope global anbox0
       valid_lft forever preferred_lft forever
    inet6 fe80::4bc:acff:fe26:32bf/64 scope link
       valid_lft forever preferred_lft forever
8: vethUzZMz6@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master anbox0 state UP group default qlen 1000
    link/ether fe:93:16:ea:22:14 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fe80::fc93:16ff:feea:2214/64 scope link
       valid_lft forever preferred_lft forever
```

</details>

The `anbox0` interface seems to be configured in ipv4 and ipv6 (both). However, the `vethUzZMz6@if2` interface only has ipv6 address. ~I'm not sure what it tells us?~ It semes like it's not an issue as it now works (after I removed the parameters I mentioned above).

The `anbox0` interface appears when I start the container and the `vethUzZMz6@if2` interface appears when I start the anbox app manager (or any anbox app I guess). I start the app with:

```
anbox launch --package=org.anbox.appmgr --component=org.anbox.appmgr.AppViewActivity
```

> There's a lot that could break in process. If none of these will help you, check your bridge configuration again just to make sure it is configured as in Arch Wiki.

Well, I used the `anbox-bridge.sh` script so I guess I can't do more than that. I also tried the `network-manager` approach, but it also failed.
